src/keydir.rs

```diff
@@ -51,7 +51,7 @@ impl KeyDir {
     pub fn get(&self, key: &[u8]) -> ErrorResult<KeyDirEntry> {
         // TODO this can just be an ok_or_else
         if !self.entries.contains_key(key) {
-            let key_str = format!("key not found: {}", String::from_utf8(key.to_vec())?);
+            let key_str = format!("key not found: {}", std::str::from_utf8(key)?);
```

Comment:
this has appeared a few times in the codebase.

so, `Slice::to_vec()` (`&[u8].to_vec()`) allocates a vec on the heap and copied all of the elements of the slice into it
and `String::from_utf8(Vec<u8>)` then moves that vec into it, verifying that its utf-8

.. but if you're just printing it out (or not needing ownership to be more clear), then you can just have str parse it and return a valid slice -- which is just borrowing from that `&[u8]`, this `str::from_utf8(&[u8])` just borrows it and validates  the utf-8 requirement.